educator/src/Dscp/Educator/Web/Types.hs

```diff
@@ -0,0 +1,79 @@
+-- | Common datatypes for educator and student HTTP APIs
+
+module Dscp.Educator.Web.Types
+       (
+         -- * Flags
+         IsFinal (..)
+       , _IsFinal
+       , HasProof (..)
+
+         -- * Responses
+       , GradeInfo (..)
+       , BlkProofInfo (..)
+
+         -- * Conversions
+       , assignmentTypeRaw
+       ) where
+
+import Control.Lens (Iso', iso, makePrisms)
+import Data.Aeson.Options (defaultOptions)
+import Data.Aeson.TH (deriveJSON)
+import Data.Time.Clock (UTCTime)
+import Servant (FromHttpApiData (..))
+
+import Dscp.Core
+import Dscp.Crypto
+import Dscp.Util.Aeson (AsHex)
+
+-- | Whether assignment is final in course.
+newtype IsFinal = IsFinal { unIsFinal :: Bool }
+    deriving (Eq, Show)
+
+makePrisms ''IsFinal
```

Comment:
Leaving the nessessity of wrapper out of scope, do we need that prism? We can unwrap it in optic chain (and then add `._Just` because thats how prisms work) and have our `Bool`, yes.

But I thought that prisms are best for distinguishing constructors.

Feel free to prove me wrong. Maybe I'm missing something here.

P.S.: why not making it `data IsFinal = IsFinal | IsNotFinal` (and adding a lens to `Bool`?), if you want to have type that is not `Bool`?

Reply:
I also think prisms are, but `IsFinal` turns to be a stronger `Iso` here, so it works for wrapping and unwrapping safely.

> why not making it data IsFinal = IsFinal | IsNotFinal (and adding a lens to Bool?)
This way I would need to define custom `ToJSON / FromJSON / FromHttpApiData` instances for servant. Not saying that's is hard, I just found a balance between not operating with blind bools and transparency for servant library.

Reply:
Then its OK.